src/torsion_profile.py

```diff
@@ -0,0 +1,191 @@
+        ## Attenuate the weights as a function of energy
+        self.set_option(tgt_opts,'attenuate','attenuate')
+        ## Energy denominator for objective function
+        self.set_option(tgt_opts,'energy_denom','energy_denom')
+        ## Set upper cutoff energy
+        self.set_option(tgt_opts,'energy_upper','energy_upper')
+        ## Read in the reference data.
+        self.read_reference_data()
+        ## Build keyword dictionaries to pass to engine.
+        engine_args = OrderedDict(list(self.OptionDict.items()) + list(options.items()))
+        del engine_args['name']
+        ## Create engine object.
+        engine_args['freeze_atoms'] = self.freeze_atoms
+        self.engine = self.engine_(target=self, mol=self.mol, **engine_args)
+
+    def read_reference_data(self):
+
+        """ Read the reference ab initio data from a file such as qdata.txt.
+
+        After reading in the information from qdata.txt, it is converted
+        into the GROMACS energy units (kind of an arbitrary choice).
+        """
+        ## Reference (QM) energies
+        self.eqm           = []
+        ## The qdata.txt file that contains the QM energies and forces
+        self.qfnm = os.path.join(self.tgtdir,"qdata.txt")
+        # Parse the qdata.txt file
+        for line in open(os.path.join(self.root,self.qfnm)):
+            sline = line.split()
+            if len(sline) == 0: continue
+            elif sline[0] == 'ENERGY':
+                self.eqm.append(float(sline[1]))
+
+        if len(self.eqm) != self.ns:
+            raise RuntimeError("Length of qdata.txt should match number of structures")
+
+        # Turn everything into arrays, convert to kcal/mol
+        self.eqm = np.array(self.eqm)
+        self.eqm *= eqcgmx / 4.184
+        # Use the minimum energy structure of the QM as reference
+        self.eqm  -= np.min(self.eqm)
+        self.smin  = np.argmin(self.eqm)
+        logger.info("Referencing all energies to the snapshot %i (minimum energy structure in QM)\n" % self.smin)
+
+        if self.attenuate:
+            # Attenuate energies by an amount proportional to their
+            # value above the minimum.
+            eqm1 = self.eqm - np.min(self.eqm)
+            denom = self.energy_denom
+            upper = self.energy_upper
+            self.wts = np.ones(self.ns)
+            for i in range(self.ns):
+                if eqm1[i] > upper:
+                    self.wts[i] = 0.0
+                elif eqm1[i] < denom:
+                    self.wts[i] = 1.0 / denom
+                else:
+                    self.wts[i] = 1.0 / np.sqrt(denom**2 + (eqm1[i]-denom)**2)
+        else:
+            self.wts = np.ones(self.ns)
+
+        # Normalize weights.
+        self.wts /= sum(self.wts)
+
+    def indicate(self):
+        title_str = "Torsion Profile: %s, Objective = % .5e, Units = kcal/mol, Angstrom" % (self.name, self.objective)
+        #QYD: This title is carefully placed to align correctly
```

Comment:
I guess this comment is copied from the `OptGeoTarget` 😀

Reply:
The rest of the comment is not wrong, so I should just change the initials. :)